pkg/controller/kubelet-config/kubelet_config_bootstrap.go

```diff
@@ -0,0 +1,175 @@
+func RunKubeletBootstrap(templateDir string, kubeletConfigs []*mcfgv1.KubeletConfig, controllerConfig *mcfgv1.ControllerConfig, features *configv1.FeatureGate, mcpPools []*mcfgv1.MachineConfigPool) ([]*mcfgv1.MachineConfig, error) {
+	var res []*mcfgv1.MachineConfig
+	generatedManagedKeys := make(map[kubeletconfigKey]string)
+	generatedMCs := make(map[string]*mcfgv1.MachineConfig)
+	userDefinedSystemReserved := make(map[string]string)
+	// Validate the KubeletConfig CR if exists
+	for _, kubeletConfig := range kubeletConfigs {
+		if kubeletConfig == nil {
+			continue
+		}
+		if err := validateUserKubeletConfig(kubeletConfig); err != nil {
+			return nil, newForgetError(err)
+		}
+	}
+
+	for _, kubeletConfig := range kubeletConfigs {
+		if kubeletConfig == nil {
+			continue
+		}
+
+		// use selector since label matching part of a KubeletConfig is not handled during the bootstrap
+		selector, err := metav1.LabelSelectorAsSelector(kubeletConfig.Spec.MachineConfigPoolSelector)
+		if err != nil {
+			return nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+
+		for _, pool := range mcpPools {
+			// If a pool with a nil or empty selector creeps in, it should match nothing, not everything.
+			// skip the pool if no matched label for kubeletconfig
+			if selector.Empty() || !selector.Matches(labels.Set(pool.Labels)) {
+				continue
+			}
+			role := pool.Name
+
+			originalKubeConfig, err := generateOriginalKubeletConfigWithFeatureGates(controllerConfig, templateDir, role, features)
+			if err != nil {
+				return nil, err
+			}
+			if kubeletConfig.Spec.TLSSecurityProfile != nil {
+				// Inject TLS Options from Spec
+				observedMinTLSVersion, observedCipherSuites := getSecurityProfileCiphers(kubeletConfig.Spec.TLSSecurityProfile)
+				originalKubeConfig.TLSMinVersion = observedMinTLSVersion
+				originalKubeConfig.TLSCipherSuites = observedCipherSuites
+			}
+
+			kubeletIgnition, logLevelIgnition, autoSizingReservedIgnition, err := generateKubeletIgnFiles(kubeletConfig, originalKubeConfig, userDefinedSystemReserved)
+			if err != nil {
+				return nil, err
+			}
+
+			tempIgnConfig := ctrlcommon.NewIgnConfig()
+			if autoSizingReservedIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *autoSizingReservedIgnition)
+			}
+			if logLevelIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *logLevelIgnition)
+			}
+			if kubeletIgnition != nil {
+				tempIgnConfig.Storage.Files = append(tempIgnConfig.Storage.Files, *kubeletIgnition)
+			}
+
+			rawIgn, err := json.Marshal(tempIgnConfig)
+			if err != nil {
+				return nil, err
+			}
+			managedKey, err := generateSuffixManagedKeyKubelet(pool, kubeletConfig, generatedManagedKeys)
+			if err != nil {
+				return nil, err
+			}
+			var (
+				mc *mcfgv1.MachineConfig
+				ok bool
+			)
+			mc, ok = generatedMCs[managedKey]
+			if !ok {
+				mc, err = ctrlcommon.MachineConfigFromIgnConfig(role, managedKey, kubeletIgnition)
```

Comment:
Does so this creates the machineconfig with only the kubeletIgnition first, and then you overwrite with the full ignition (rawIgn) right below this? If so why go through such a hoop?

Reply:
Removed this logic. Thanks for the reminder, I used `generatedMCs` in my implementation that is not necessary now.